src/socket.cr

```diff
@@ -31,110 +31,310 @@ class Socket < IO::FileDescriptor
     INET6  = LibC::AF_INET6
   end
 
-  struct IPAddress
-    getter family : Family
-    getter address : String
-    getter port : UInt16
-
-    def initialize(@family : Family, @address : String, port : Int)
-      if family != Family::INET && family != Family::INET6
-        raise ArgumentError.new("Unsupported address family")
-      end
+  # :nodoc:
+  SOMAXCONN = 128
+
+  property family : Family
+  property type : Type
+  property protocol : Protocol
+
+  # Creates a TCP socket. Consider using `TCPSocket` or `TCPServer` unless you
+  # need full control over the socket.
+  def self.tcp(family : Family, blocking = false)
+    new(family, Type::STREAM, Protocol::TCP, blocking)
+  end
+
+  # Creates an UDP socket. Consider using `UDPSocket` unless you need full
+  # control over the socket.
+  def self.udp(family : Family, blocking = false)
+    new(family, Type::DGRAM, Protocol::UDP, blocking)
+  end
 
-      @port = port.to_u16
+  # Creates an UNIX socket. Consider using `UNIXSocket` or `UNIXServer` unless
+  # you need full control over the socket.
+  def self.unix(type : Type = Type::Stream, blocking = false)
+    new(Family::UNIX, type, blocking: blocking)
+  end
+
+  def initialize(@family, @type, @protocol = Protocol::IP, blocking = false)
+    fd = LibC.socket(family, type, protocol)
+    raise Errno.new("failed to create socket:") if fd == -1
+    init_close_on_exec(fd)
+    super(fd, blocking)
+    self.sync = true
+  end
+
+  protected def initialize(fd : Int32, @family, @type, @protocol = Protocol::IP)
+    init_close_on_exec(fd)
+    super fd, blocking: false
+    self.sync = true
+  end
+
+  # Force opened sockets to be closed on `exec(2)`. Only for platforms that don't
+  # support `SOCK_CLOEXEC` (e.g., Darwin).
+  protected def init_close_on_exec(fd : Int32)
+    {% unless LibC.constants.includes?("SOCK_CLOEXEC".id) %}
```

Comment:
You can also use `LibC.has_constant?("SOCK_CLOEXEC")` (shorter plus probably a bit faster.) Not sure it works on the old compiler without `.id`, so it's probably better to leave the `.id` for now